contracts/core/UniversalLiquidatorRegistry.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+// imported contracts and libraries
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+// interfaces
+import "../interface/IUniversalLiquidatorRegistry.sol";
+
+// libraries
+import "../libraries/DataTypes.sol";
+import "../libraries/Errors.sol";
+
+// constants and types
+import {ULRegistryStorage} from "./storage/ULRegistry.sol";
+
+contract UniversalLiquidatorRegistry is
+    Ownable,
+    IUniversalLiquidatorRegistry,
+    ULRegistryStorage
+{
+    function getPath(
+        address _sellToken,
+        address _buyToken
+    ) public view override returns (DataTypes.SwapInfo[] memory) {
+        if (paths[_sellToken][_buyToken].dex != bytes32(0)) {
+            DataTypes.SwapInfo[] memory retPaths = new DataTypes.SwapInfo[](1);
+            retPaths[0] = DataTypes.SwapInfo(
+                dexesInfo[paths[_sellToken][_buyToken].dex],
+                paths[_sellToken][_buyToken].paths
+            );
+            return retPaths;
+        }
+
+        for (uint256 idx; idx < intermediateTokens.length; ) {
+            if (
+                paths[_sellToken][intermediateTokens[idx]].dex != bytes32(0) &&
+                paths[intermediateTokens[idx]][_buyToken].dex != bytes32(0)
+            ) {
+                // found the intermediateToken and intermediateDex
+                DataTypes.SwapInfo[] memory retPaths = new DataTypes.SwapInfo[](
+                    2
+                );
+                retPaths[0] = DataTypes.SwapInfo(
+                    dexesInfo[paths[_sellToken][intermediateTokens[idx]].dex],
+                    paths[_sellToken][intermediateTokens[idx]].paths
+                );
+                retPaths[1] = DataTypes.SwapInfo(
+                    dexesInfo[paths[intermediateTokens[idx]][_buyToken].dex],
+                    paths[intermediateTokens[idx]][_buyToken].paths
+                );
+                return retPaths;
+            }
+            unchecked {
+                ++idx;
+            }
+        }
+        revert("Liquidation path is not set");
+    }
+
+    function setPath(
+        address _sellToken,
+        address _buyToken,
+        bytes32 _dex,
+        address[] memory _paths
+    ) external override onlyOwner {
+        // path could also be an empty array
+        if (_sellToken != _paths[0]) revert Errors.InvalidPathsStart();
+        if (_buyToken != _paths[_paths.length - 1])
+            revert Errors.InvalidPathsEnd();
+
+        // path can also be empty
+        paths[_sellToken][_buyToken] = DataTypes.PathInfo(_dex, _paths);
+    }
+
+    function addIntermediateToken(address _token) public override onlyOwner {

```

Comment:
We might also want to add a function to remove an intermediate token from the list.

Reply:
Fixed

Reply:
Should we change it to a "Set", would that be easier to maintain?